File: src/python/busca_binaria.py
""" Implementation of the binary search algorithm with recursion """


def binary_search(value, vector, left, right):
    """
    Implementation of a binary search algorithm with recursion.

    Arguments:
    value: Any. Value to be searched for in the list
    vector: list. ordered list in which the value will be searched
    left: Any. Initial value of the half sought
    right: Any. Final value of the half sought

    Returns the index of the value in "vector" or -1 if it does not exist in it.
    """
    middle = int((left + right) / 2)

    if left <= right:
        if value > vector[middle]:
            left = middle + 1
            return binary_search(value, vector, left, right)
        elif value < vector[middle]:
            right = middle - 1
            return binary_search(value, vector, left, right)
        return middle
    return -1


list = [0, 1, 3, 5, 6, 7, 8, 9, 10, 11, 12]
print(binary_search(12, list, 0, len(list) - 1))

File: src/go/busca_binaria.go
package main

import "fmt"

// Recursive Approach
func BinarySearch(seq []int, element, start, end int) int {
    index := (start + end) / 2 // Calculate the middle index

    // If the element is found
    if index < len(seq) && seq[index] == element {
        return index
    }

    // If the element is not found
    if start != end {
        // If the element is greater than the middle element
        if seq[index] < element {
            return BinarySearch(seq, element, index+1, end) // Search in the right side
        }

        // If the element is less than the middle element
        return BinarySearch(seq, element, start, index-1) // Search in the left side
    }

    return -1 // Element not found in the slice
}

func main() {
    value := 7
    slice := []int{1, 2, 3, 4, 5, 6, 7, 8, 9}
    fmt.Println("Slice:", slice)
    index := BinarySearch(slice, value, 0, len(slice))
    fmt.Println("Value", value, "found in index:", index)
}
